contracts/AuctionBase.sol

```diff
@@ -129,59 +128,101 @@ contract AuctionBase is Pausable {
       highestBidder: address(0),
       cancelled: false
     });
-
     uint256 newAuctionId = auctions.push(_auction) - 1;
 
     // Add auction index to nftToTokenIdToAuctionId mapping
     nftToTokenIdToAuctionId[_nftAddress][_tokenId] = newAuctionId;
 
-    // Emit AuctionCreated event
     AuctionCreated(newAuctionId, _nftAddress, _tokenId);
   }
 
+  // @dev Implements a simplified English auction
+  // Lets msg.sender bid highestBid + bidIncrement and stores bids in fundsByBidder
+  // TODO: Look into the experience of bidding in an English Auction asyncronously
   function bid(uint256 _auctionId)
     external
     payable
     whenNotPaused
     statusIs(AuctionStatus.Active, _auctionId)
+    returns (bool success)
   {
     require(msg.value > 0);
 
-    // Get auction from _id
     Auction storage auction = auctions[_auctionId];
 
-    // Set newBid
-    uint256 newBid;
-    if (auction.highestBidder == msg.sender) {
-      newBid = auction.highestBid + msg.value;
-    } else {
-      newBid = auction.allowed[msg.sender] + msg.value;
-    }
-
-    // Require newBid be more than highestBid
-    require(newBid > auction.highestBid);
+    // Require newBid be greater than or equal to highestBid + bidIncrement
+    uint256 newBid = auction.fundsByBidder[msg.sender] + msg.value;
+    require(newBid >= auction.highestBid + auction.bidIncrement);
 
-    // Update allowed mapping
-    if (auction.highestBidder != msg.sender) {
-      auction.allowed[auction.highestBidder] = auction.highestBid;
-      auction.allowed[msg.sender] = 0;
-    }
-    auction.highestBidder = msg.sender;
+    // Update fundsByBidder mapping
     auction.highestBid = newBid;
+    auction.highestBidder = msg.sender;
+    auction.fundsByBidder[auction.highestBidder] = newBid;
 
     // Emit BidCreated event
     BidCreated(_auctionId, auction.nftAddress, auction.tokenId, msg.sender, newBid);
+    return true;
+  }
+
+  // @dev Allow people to withdraw their balances or the NFT
+  function withdrawBalance(uint256 _auctionId) external returns (bool success) {
+    AuctionStatus _status = _getAuctionStatus(_auctionId);
+
+    // Don't let people withdraw before the auction is cancelled or completed
+    require(_status == AuctionStatus.Cancelled || _status == AuctionStatus.Completed);
```

Comment:
Someone should be able to withdraw their funds if not the highest bidder, while auction is still accepting bids

Reply:
Good call

Reply:
Could make the logic a little weird, but will look into it